students/stefan_lund/Lesson_4/mailroom_part2.py

```diff
@@ -77,296 +80,314 @@ def send_letters_to_everyone(menu, file_name):
-
-    return False
-
-def letter_templates(version):
-
-
+        letter_file = donor + "__" + today[0] + "_" + today[1] + "_" + today[2] + "_b"
+        pth = os.path.join(os.getcwd(), "letter_to_everyone")
+        if not os.path.exists(pth):
+            os.makedirs(pth)
+        destination = os.path.join(pth, letter_file)
+
+        try:
+            with open(destination, 'w') as outfile:
+                for line in letter:
+                    outfile.write(line)
+        except:
+            print("something went wrong, read")
 
-    n = "\n"
-    t = "\t"
-    s = " "
-
-    # letter 1, date: todays date, name, amount: amount of donation registered today
-    l1 = "Seattle, WA {date}" + n * 3 + t + "Dear {name}," + n * 2
-    l2 = ((t + s) * 1 + "Thank you for your very "
-         "kind donation of ${amount}." + n * 2)
-    l3 = (t + s) * 1 + "It will be put to very good use." + n * 2
-    l4 = (t + s) * 5 + "Sincerely," + n * 2
-    l5 = (t + s) * 6 + "-The Team"
-    letter1 = l1 + l2 + l3 + l4 + l5
-
-
-    # letter 2, date: todays date, name, amount: total amount donated
-    day = (t + s) * 8 + "{date}" + n * 6
-    l1 = "Dear {name}," + n * 2
-    l2 = (t + s) * 1 + "Your very generous donation of ${amount} \
-         this year is very much appreciated." + n * 2
-    l3 = (t + s) * 1 + "It will be put to very good use." + n * 2
-    l4 = (t + s) * 5 + "Sincerely," + n * 2
-    l5 = (t + s) * 6 + "-The Team"
-    letter2 = day + l1 + l2 + l3 + l4 + l5
+    option = "start"
+    finished = False
+    return finished, option
 
-    version_dict = {"1":letter1, "2":letter1, "3":"under construction"}
 
-    return version_dict[version]
-
-def todays_date():
-    """
-        returns a list containing year, month, day in str format
-    """
-    today = datetime.date.today()
-    year, month, day = str(today.year), str(today.month), str(today.day)
-    return [year, month, day]
+# "thank you" menu functions
 
-def get_names_from(in_file):
+def main_menu():
     """
-        in_file: file name for file in cwd where data is stored
-        returns: list of recorded names
+    displays the "start" or the "thank you" menu according to current option
+    default when script starts is "start" menu
+    variable "finished" is False for all choices except "Quit" which returns
+    "finished" as True
     """
-    data = get_data_from(in_file)
-    name_list = []
-    for name in data:
-        name_list.append(name)
-
-    return name_list
+    global menu
+    option = "start"
+    finished = False
+    while not finished:
+        sub_menu = menu[option]
+        print("1.  main_menu, ", finished, option)
+        finished, option = ask_questions(sub_menu)
+
+        # when "Quit returns "finished as True the while loop ends but script
+        # crashes, the solution with if and break was added
+        if finished:
+            print("2.  main_menu, ", finished, option)
+            store_data()
+            break

```

Comment:
How do I avoid using eval?
I know, my fault. The assignment never asked for this so I understand if you find that now is not the moment for this voluntarily added problem. 
Just in case it's ok, my thinking was that the menu is coming from somewhere outside the function using it. I declared it a global variable here. Problem is to accomplish this, the function is of a string type. How is the string name of the function converted into an executable function? From "my_function" to my_function?

Reply:
my example and the link above show how.. but let try to explain another way:
```
>>> def fun1():
...     print('in fun1')
...
>>> def fun2():
...     print('in fun2')
...
>>> switch_dict = {1: fun1, 2: fun2}
>>> switch_dict[1]()
in fun1
>>>
```
above I mapped two keys to two different function objects, does the above code make sense? try it in the python shell.

Reply:
remember that your switch dict would be "close" to your main entry point, as it is an integral part of that code block - so it wouldn't really be "coming from somewhere outside the function" 

please let me know if you have any further questions 🙂

Reply:
Works great. 

Here the switch_dict is not within the function using it. The function name has to be a string or the error 	E0602 Undefined variable 'Send_a_Thank_You' is thrown.

Can I get around it without using eval?

Reply:
```
  1 def fun1():
  2     print("in fun1")
  3
  4
  5 def main():
  6     switch_dict[1]()  # this would wrapped in the while True block in the mailroom
  7
  8
  9
 10 if __name__ == "__main__":
 11     switch_dict = {1: fun1}
 12     main()
 13
```

^^ above is a full example - you need to define your switch dict in the main block and you will not have resolution problems.